### PR TITLE
DM-5292: Add queue support

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

File: include/lsst/ctrl/events/Event.h [hunk 1]
```diff
@@ -42,7 +42,8 @@
 #include <iostream>
 #include <set>
 
-#include "lsst/utils/Utils.h"
+#include "boost/shared_ptr.hpp"
```
Comment (timj) [inline include/lsst/ctrl/events/Event.h]: Can we group the boost includes together?

Comment (srp3rd) [inline include/lsst/ctrl/events/Event.h]: Will do

File: include/lsst/ctrl/events/EventSystem.h [hunk 2]
```diff
@@ -85,8 +88,16 @@ class EventSystem {
...
+     * @param hostName the location of the message broker to use
+     * @param queueName the queue to transmit events to
+     * @param hostPort the port where the broker can be reached
+     * @throws lsst::pex::exceptions::RuntimeError if queue is already registered
+     */
+    void createEnqueuer(std::string const& hostName, std::string const& queueName, int hostPort = EventBroker::DEFAULTHOSTPORT);
 
-    /** 
+    /**
      * @brief create an EventReceiver which will receive message
```
Comment (r-owen) [inline include/lsst/ctrl/events/EventSystem.h:99]: Should this say "receive messages"?

File: include/lsst/ctrl/events/EventSystem.h [hunk 3]
```diff
@@ -95,7 +106,16 @@ class EventSystem {
...
 
-    /** 
+    /**
+     * @brief create a EventDequeuer which will receive message
+     * @param hostName the location of the message broker to use
+     * @param queueName the queue to receive messages from
+     * @param hostPort the port where the broker can be reached
+     * @throws lsst::pex::exceptions::RuntimeError if topic is already registered
+     */
+    void createDequeuer(std::string const& hostName, std::string const& queueName, int hostPort = EventBroker::DEFAULTHOSTPORT);
```
Comment (r-owen) [inline include/lsst/ctrl/events/EventSystem.h:114]: Please move this above the definition of `createReceiver`, in order to keep the various overloads of each function together.

File: include/lsst/ctrl/events/EventTransmitter.h [hunk 4]
```diff
@@ -41,41 +41,38 @@
...
+
+    typedef boost::shared_ptr<EventTransmitter> Ptr;
+
     /** 
-     * @brief Transmits events to the specified host and topic
+     * @brief Transmits events to the specified host and destination
      *
      * @param hostName the machine hosting the message broker
-     * @param topicName the topic to transmit events to
+     * @param destinationName the topic to transmit events to
```
Comment (r-owen) [inline include/lsst/ctrl/events/EventTransmitter.h:65]: I'm a bit confused by the interchangeable use of topic and destination. Should the help for destinationName really be using "topic"? Perhaps a few more words would help?

File: include/lsst/ctrl/events/Transmitter.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    // JMS session
+    cms::Session* _session;
+
+    void init( const std::string& hostName, const std::string& destinationName, bool createQueue, int port);
+
+private:
+
+    // Connection to JMS broker
+    cms::Connection* _connection;
```
Comment (r-owen) [inline include/lsst/ctrl/events/Transmitter.h:100]: Please explain what "JMS" is, either here or in the class doc or...?

File: include/lsst/ctrl/events/Transmitter.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * @brief Publish an Event to this object's topic
+     * @param event an Event to publish
+     */
+    void publishEvent(Event& event);
+
+    /**
+     * @brief get the destination property name
+     */
```
Comment (r-owen) [inline include/lsst/ctrl/events/Transmitter.h:80]: Please explain what a "destination property name" is -- here and in Receiver.h

File: include/lsst/ctrl/events/Transmitter.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+     * @brief Publish an Event to this object's topic
+     * @param event an Event to publish
+     */
+    void publishEvent(Event& event);
+
+    /**
+     * @brief get the destination property name
+     * @note This is the TYPE of the destination we're using, either a TOPIC or a QUEUE
+     */
+    virtual std::string getDestinationPropertyName() { return NULL; }
```
Comment (andy-slac) [inline include/lsst/ctrl/events/Transmitter.h:81]: Sorry, I realize it's too late for comments, but I still have to say it :smile: 
I hope this method is never called (overridden in all subclasses), constructing an `std::string` from zero pointer will very likely result in a crash. If you want an empty string use `""` or explicit `std::string()` (I prefer latter).

File: include/lsst/ctrl/events/Receiver.h [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+     * @param timeout the length of time to wait in milliseconds; value of -1 waits indefinately.
+     * @return an Event
+     */
+    PTR(Event) receiveEvent(long timeout);
+
+    /**
+     * @brief get the destination property name
+     * @note This is the TYPE of the destination we're using, either a TOPIC or a QUEUE
+     */
+    virtual std::string getDestinationPropertyName() { return NULL; }
```
Comment (andy-slac) [inline include/lsst/ctrl/events/Receiver.h:89]: And same comment here.

File: include/lsst/ctrl/events/Transmitter.h (hunk 7 above)
Comment (r-owen) [inline include/lsst/ctrl/events/Transmitter.h:81]: Good catch. I suggest you file a ticket so it can get fixed.

Comment (andy-slac) [inline include/lsst/ctrl/events/Transmitter.h:81]: OK, done, DM-7009 assigned to @srp3rd.

**Pull request merged**